src/Std/Data/DHashMap/Internal/RawLemmas.lean

```diff
@@ -832,13 +832,196 @@ theorem contains_keys [EquivBEq α] [LawfulHashable α] (h : m.1.WF) {k : α} :
+
+
+theorem get?_insertList_not_mem_list [LawfulBEq α] [LawfulHashable α] {l: List ((a:α) × (β a))} {k k': α} {mem: ¬ k ∈ (l.map (Sigma.fst))} {k_eq: k == k'} (h: m.1.WF):
+    (m.insertList l).get? k' = m.get? k' := by
+  simp_to_model using getValueCast?_insertList_not_toInsert_mem
+
+theorem get!_insertList_mem_list [LawfulBEq α] [LawfulHashable α]  {l: List ((a:α) × (β a))} {k k': α} {v: β k}[Inhabited (β k')] {distinct: List.Pairwise (fun a b => ! a.1 == b.1) l} {k_eq: k == k'} {mem: ⟨k,v⟩ ∈ l} (h: m.1.WF):
+    (m.insertList l).get! k' = cast (by congr; apply LawfulBEq.eq_of_beq k_eq) v := by
+  simp_to_model using getValueCast!_insertList_toInsert_mem
+
+theorem get!_insertList_not_mem_list [LawfulBEq α] [LawfulHashable α] {l: List ((a:α) × (β a))} {k k': α} [Inhabited (β k')] {mem: ¬ k ∈ (l.map (Sigma.fst))} {k_eq: k == k'} (h: m.1.WF):
+    (m.insertList l).get! k' = m.get! k' := by
+  simp_to_model using getValueCast!_insertList_not_toInsert_mem
+
+theorem getD_insertList_mem_list [LawfulBEq α] [LawfulHashable α]  {l: List ((a:α) × (β a))} {k k': α} {v: β k} {fallback: β k'} {distinct: List.Pairwise (fun a b => ! a.1 == b.1) l} {k_eq: k == k'} {mem: ⟨k,v⟩ ∈ l} (h: m.1.WF):
+    (m.insertList l).getD k' fallback = cast (by congr; apply LawfulBEq.eq_of_beq k_eq) v := by
+  simp_to_model using getValueCastD_insertList_toInsert_mem
+
+theorem getD_insertList_not_mem_list [LawfulBEq α] [LawfulHashable α] {l: List ((a:α) × (β a))} {k k': α} {fallback: β k'} {mem: ¬ k ∈ (l.map (Sigma.fst))} {k_eq: k == k'} (h: m.1.WF):
+    (m.insertList l).getD k' fallback = m.getD k' fallback := by
+  simp_to_model using getValueCastD_insertList_not_toInsert_mem
+
+
+theorem get_insertList_mem_list [LawfulBEq α] [LawfulHashable α] {l: List ((a:α) × (β a))} {k k': α} {v: β k} {distinct: List.Pairwise (fun a b => ! a.1 == b.1) l} {k_eq: k == k'} {mem: ⟨k,v⟩ ∈ l} (h: m.1.WF):
+    (m.insertList l).get k' (contains_insertList_of_mem_list _ h k_eq mem)= cast (by congr; apply LawfulBEq.eq_of_beq k_eq ) v := by
+  simp_to_model using getValueCast_insertList_toInsert_mem
+
+theorem contains_of_beq [EquivBEq α][LawfulHashable α] {k k': α} (k_eq: k == k') (h: m.1.WF):
+    m.contains k = true → m.contains k' = true := by
+  simp_to_model
+  intro h'
+  apply containsKey_of_beq h' k_eq
+
+theorem get_insertList_not_mem_list [LawfulBEq α] [LawfulHashable α] {l: List ((a:α) × (β a))} {k k': α} {k_eq: k == k'} {mem_list: ¬ k ∈ (l.map (Sigma.fst))} (h: m.1.WF):
+    (h': m.contains k = true) → (m.insertList l).get k' (contains_insertList_of_contains_map m h k_eq h') = m.get k' (contains_of_beq m k_eq h h') := by
+  simp_to_model using getValueCast_insertList_not_toInsert_mem
+
+theorem getKey?_insertList [EquivBEq α] [LawfulHashable α] (h : m.1.WF) {l : List ((a:α) × (β a))} {k : α} :
+    (m.insertList l).getKey? k =
+      if (l.map Sigma.fst).contains k then (l.map Sigma.fst).reverse.find? (fun a => k == a) else m.getKey? k := by
```

Comment:
Would it be preferred to split this into two theorems rather than using the if? (same below)